### DISCUSSION TITLE
NotNullIfNotNull is not recognized inside a generic method body.

### DISCUSSION THREAD

**Discussion by manfred-brands** [dotnet/roslyn#45612] [General]
**Version Used**: 
VS 2019 16.6.2 
C# 8 with nullability enabled.

**Steps to Reproduce**:
```
        [return: NotNullIfNotNull("defaultValue")]
        private static T Find<T>([AllowNull] T defaultValue)
        {
            return defaultValue; // CS8603
        }

        [return: NotNullIfNotNull("defaultValue")]
        private static T FindOrCreate<T>([AllowNull] T defaultValue)
        {
            return Find(defaultValue);
        }
```

**Expected Behavior**:

No CS8603 raised at the return defaultValue statement.

**Actual Behavior**:

CS8603 is raised at the return defaultValue statement.
Adding [return: MaybeNull] removes the CS8603 in the first function, but then raises it in the second.
It is related to [36039](https://github.com/dotnet/roslyn/issues/36039) which fixed it for MaybeNull.

**Comment by RikkiGibson** [comment]
My interpretation here is that while you have `[return: NotNullIfNotNull]`, these methods are still not permitted to return a maybe-null value when T is a non-nullable reference type.

There are two possible solutions to this:

1. You could add `[return: MaybeNull]` to both methods, since as written, both might return a maybe-null value when T is a non-nullable reference type.

2. You could delete the `[AllowNull]` attribute from the parameters, which would make it so maybe-null values are only acceptable when T is a nullable reference type.

**Reply by manfred-brands** [reply]
With C#9 I can actually declared the return as `T?` solving the problem.
